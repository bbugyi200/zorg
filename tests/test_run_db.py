"""Tests the zorg project's 'db' CLI command."""

from __future__ import annotations

from pathlib import Path
import sqlite3

from _pytest.tmpdir import TempPathFactory
from pytest import fixture
from syrupy.assertion import SnapshotAssertion as Snapshot

from . import common as c


@fixture(name="sql_cursor", scope="session")
def sql_cursor_fixture(
    main: c.MainType, tmp_path_factory: TempPathFactory, zettel_dir: Path
) -> sqlite3.Cursor:
    """Connects to the zorg DB created by 'zorg db create'."""
    tmp_path = tmp_path_factory.getbasetemp()
    db_path_str = f"{tmp_path}/zorg.db"
    exit_code = main(
        "--dir",
        str(zettel_dir),
        "db",
        "create",
        database_url=f"sqlite:///{db_path_str}",
    )
    assert exit_code == 0
    conn = sqlite3.connect(db_path_str)
    return conn.cursor()


def test_run_db_create__table_names(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check the SQL table names generated by running 'zorg db create'."""
    sql_cursor.execute("SELECT name FROM sqlite_master WHERE type = 'table'")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__file_count(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check the number of *.zo files indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT COUNT(*) FROM zorgfile")
    assert snapshot == sql_cursor.fetchone()[0]


def test_run_db_create__note_count(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check the number of zorg notes indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT COUNT(*) FROM zorgnote")
    assert snapshot == sql_cursor.fetchone()[0]


def test_run_db_create__todo_count(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check the number of zorg todos indexed by running 'zorg db create'."""
    sql_cursor.execute(
        "SELECT COUNT(*) FROM zorgnote WHERE todo_status IS NOT NULL"
    )
    assert snapshot == sql_cursor.fetchone()[0]


def test_run_db_create__projects(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what projects are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT name FROM project")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__areas(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what areas are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT name FROM area")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__people(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what people are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT name FROM person")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__contexts(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what contexts are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT name FROM context")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__properties(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what properties are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT name FROM property")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__create_dates(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what create dates are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT DISTINCT create_date FROM zorgnote")
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__todo_priorities(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what todo priorities are indexed by running 'zorg db create'."""
    sql_cursor.execute(
        "SELECT DISTINCT todo_priority FROM zorgnote WHERE todo_priority IS"
        " NOT NULL"
    )
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__todo_statuses(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what todo statuses are indexed by running 'zorg db create'."""
    sql_cursor.execute(
        "SELECT DISTINCT todo_status FROM zorgnote WHERE todo_status IS NOT"
        " NULL"
    )
    assert snapshot == sorted(row[0] for row in sql_cursor.fetchall())


def test_run_db_create__zids(
    sql_cursor: sqlite3.Cursor, snapshot: Snapshot
) -> None:
    """Check what create dates are indexed by running 'zorg db create'."""
    sql_cursor.execute("SELECT DISTINCT zid FROM zorgnote")
    assert snapshot == sorted(
        row[0] for row in sql_cursor.fetchall() if row[0]
    )
