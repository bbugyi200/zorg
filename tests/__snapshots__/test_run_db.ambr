# serializer version: 1
# name: test_run_db_create__areas[YAMLConfigFile]
  list([
    'gtd',
    'own',
    'tags',
    'thoughts',
  ])
# ---
# name: test_run_db_create__contexts[YAMLConfigFile]
  list([
    'due',
    'dzwk',
    'fat_pig',
    'pomodoro',
    'puke',
  ])
# ---
# name: test_run_db_create__create_dates[YAMLConfigFile]
  list([
    '2000-01-03',
    '2024-03-13',
    '2024-03-23',
    '2024-03-30',
    '2024-04-08',
    '2030-12-31',
  ])
# ---
# name: test_run_db_create__file_count[YAMLConfigFile]
  10
# ---
# name: test_run_db_create__note_count[YAMLConfigFile]
  135
# ---
# name: test_run_db_create__people[YAMLConfigFile]
  list([
    'clyzajrm',
    'pig',
  ])
# ---
# name: test_run_db_create__projects[YAMLConfigFile]
  list([
    'arms',
    'bad',
    'bazbuz',
    'bmlakcr',
    'bqf',
    'cmeq',
    'cpv_mociy',
    'fgifr',
    'foobar',
    'ipyt',
    'judo',
    'jws_pkaft',
    'laundry',
    'lozmhqjd',
    'okwz',
    'owwnfb',
    'qbsum',
    'rgw_sxtvx',
    'saxy',
    'to_the_bone',
    'uahi',
    'vd_clxi',
    'vennc',
    'vfntwi',
    'wjybwa',
    'xoao',
    'zorg',
    'zv_jzdd_djqa',
  ])
# ---
# name: test_run_db_create__properties[YAMLConfigFile]
  list([
    'due',
    'key',
    'p',
  ])
# ---
# name: test_run_db_create__table_names[YAMLConfigFile]
  list([
    'area',
    'arealink',
    'context',
    'contextlink',
    'person',
    'personlink',
    'project',
    'projectlink',
    'property',
    'propertylink',
    'zorgfile',
    'zorgfilelink',
    'zorgnote',
  ])
# ---
# name: test_run_db_create__todo_count[YAMLConfigFile]
  12
# ---
# name: test_run_db_create__todo_priorities[YAMLConfigFile]
  list([
    'A',
    'B',
    'C',
  ])
# ---
# name: test_run_db_create__todo_statuses[YAMLConfigFile]
  list([
    'OPEN_TODO',
  ])
# ---
# name: test_run_db_create__zids[YAMLConfigFile]
  list([
    '000103#00',
    '000103#01',
    '000103#02',
    '000103#03',
    '000103#04',
    '000103#05',
    '000103#06',
    '000103#07',
    '000103#08',
    '000103#09',
    '000103#0A',
    '000103#0B',
    '000103#0C',
    '000103#0D',
    '000103#0E',
    '000103#0F',
    '000103#0G',
    '000103#0H',
    '000103#0J',
    '000103#0K',
    '000103#0L',
    '000103#0M',
    '000103#0N',
    '000103#0P',
    '000103#0R',
    '000103#0T',
    '000103#0U',
    '000103#0V',
    '000103#0W',
    '000103#0X',
    '000103#0Y',
    '000103#0Z',
    '000103#0a',
    '000103#0b',
    '000103#0c',
    '000103#0d',
    '000103#0e',
    '000103#0f',
    '000103#0h',
    '000103#0k',
    '000103#0m',
    '000103#0n',
    '000103#0o',
    '000103#0r',
    '000103#0s',
    '000103#0t',
    '000103#0u',
    '000103#0v',
    '000103#0w',
    '000103#0x',
    '000103#0z',
    '000103#10',
    '000103#11',
    '000103#12',
    '000103#13',
    '000103#14',
    '000103#15',
    '000103#16',
    '000103#17',
    '000103#18',
    '000103#19',
    '000103#1A',
    '000103#1B',
    '000103#1C',
    '000103#1D',
    '000103#1E',
    '000103#1F',
    '000103#1G',
    '000103#1H',
    '000103#1J',
    '000103#1K',
    '000103#1L',
    '000103#1M',
    '000103#1N',
    '000103#1P',
    '000103#1R',
    '000103#1T',
    '000103#1U',
    '000103#1V',
    '000103#1W',
    '000103#1X',
    '000103#1Y',
    '000103#1Z',
    '000103#1a',
    '000103#1b',
    '000103#1c',
    '000103#1d',
    '000103#1e',
    '000103#1f',
    '000103#1h',
    '000103#1k',
    '000103#1m',
    '000103#1n',
    '000103#1o',
    '000103#1r',
    '000103#1s',
    '000103#1t',
    '000103#1u',
    '000103#1v',
    '000103#1w',
    '000103#1x',
    '000103#1z',
    '000103#20',
    '000103#21',
    '000103#22',
    '000103#23',
    '000103#24',
    '000103#25',
    '000103#26',
    '000103#27',
    '240313#00',
    '240313#01',
    '240313#02',
    '240323#00',
    '240323#01',
    '240323#02',
    '240323#03',
    '240323#04',
    '240323#05',
    '240323#06',
    '240323#07',
    '240323#08',
    '240323#09',
    '240323#0A',
    '240323#0B',
    '240323#0C',
    '240323#0D',
    '240323#0E',
    '240323#0F',
    '240323#0G',
    '240323#0H',
    '240330#00',
    '240330#01',
    '240408#0X',
    '301231#X0',
  ])
# ---
